# Live Challenge 1

## Perkenalan

Judan Syamsul Hadad / FTDS009

Notebook ini dibuat sebagai pemenuhan penilaian kompetensi dari rangkaian <u>**Full Time Data Science (FTDS) Bootcamp Batch 9**</u> yang diadakan oleh Hacktiv8 dimana penulis atas nama **Judan Syamsul Hadad** atau dengan handlename **Riezn** merupakan salah satu pesertanya.

## Soal Nomor 1

1. Menggunakan Pandas query, ambil data kasus kumulatif Covid-19 tanggal 20 Januari 2022 untuk negara-negara dengan latitude di bawah 0. Negara mana yang paling tinggi jumlah kumulatif kasus Covidnya per tanggal tersebut? Jika bukan Indonesia, bandingkan jumlah kumulatif kasus negara tersebut dengan Indonesia per tanggal yang sama!

### Import Library

In [209]:
import numpy as np
import pandas as pd

### Data Loading

In [210]:
df = pd.read_csv("https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_confirmed_global.csv") #read csv dari link

Lihat isi dataset

In [211]:
df.head() #cek lima baris teratas

Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  2/10/22  2/11/22  2/12/22  \
0        0        0        0        0  ...   169448   169940   170152   
1        0        0        0        0  ...   267020   267020   267551   
2        0        0        0        0  ...   260191   260723   261226   
3        0        0        0        0  ...    37074    37140    37140   
4        0        0        0        0  ...    98474    98501    98514   

   2/13/22  2/14/22  2/15/22  2/16/22  2/17/22  2/18/22  2/19/22  
0   170604   171246   171422   171519   171673   171857   171931  
1   268008   268304   268491   268940   269301   269601   269904  
2   261752   262165   262570   262994   263369   263685   263936  
3    37140    37277    37361    37452    37522    37589    37589  
4    98514    98514    98555    98568    98585    98605    98617  

[5 rows x 764 columns]

### Answer

Data kasus Covid-19 tanggal 20 Januari 2022 di negara dengan Latitude di bawah 0

In [212]:
df[
    (df['Lat']<0) #query untuk baris dengan nilai Lat di bawah 0
][['Country/Region','Lat','1/20/22']] #menampilkan kolom Country/Region, Lat, dan tanggal 20 Januari 2022

Country/Region        Lat   1/20/22
4                 Angola -11.202700     95220
5             Antarctica -71.949900        11
7              Argentina -38.416100   7576335
9              Australia -35.473500     28472
10             Australia -33.868800    919876
11             Australia -12.463400      8219
12             Australia -27.469800    298367
13             Australia -34.928500     89043
14             Australia -42.882100     23515
15             Australia -37.813600    717692
16             Australia -31.950500      1338
28               Bolivia -16.290200    795087
30              Botswana -22.328500    243946
31                Brazil -14.235000  23595178
36               Burundi  -3.373100     36868
58                 Chile -35.675100   1916522
94               Comoros -11.645500      7784
95   Congo (Brazzaville)  -0.228000     23244
96      Congo (Kinshasa)  -4.038300     84198
110              Ecuador  -1.831200    658045
116             Eswatini -26.522500     68003
118                 Fiji -17.713400     60509
121               France -17.679700     47275
124               France -12.827500     35189
125               France -20.904305     14854
126               France -21.115100    133617
130               France -14.293800       454
132                Gabon  -0.803700     45663
149            Indonesia  -0.789300   4277644
159                Kenya  -0.023600    319379
160             Kiribati  -3.370400        37
168              Lesotho -29.610000     31899
175           Madagascar -18.766947     55827
176               Malawi -13.254300     83403
183            Mauritius -20.348404     70094
191           Mozambique -18.665695    221432
192              Namibia -22.957600    154897
199          New Zealand -21.236700         2
200          New Zealand -40.900600     15401
210     Papua New Guinea  -6.314993     36446
211             Paraguay -23.442500    516555
212                 Peru  -9.190000   2780049
219               Rwanda  -1.940300    127227
223                Samoa -13.759000        13
229           Seychelles  -4.679600     34367
234      Solomon Islands  -9.645700        81
236         South Africa -30.559500   3572860
248             Tanzania  -6.369028     32393
250          Timor-Leste  -8.874217     19866
252                Tonga -21.179000         1
265       United Kingdom -51.796300        85
271       United Kingdom  -7.946700         4
274              Uruguay -32.522800    556163
276              Vanuatu -15.376700         7
282               Zambia -13.133897    299971
283             Zimbabwe -19.015438    227552

Sekarang cek jumlah kasus kumulatif terbanyak di dunia pada query di atas

In [213]:
df[
    (df['Lat']<0) #query untuk baris dengan nilai Lat di bawah 0
].groupby(['Country/Region'])['1/20/22'].max().max() #grouping berdasarkan negara agar nilai kasus max per negara dapat diketahui, lalu mengambil nilai max dari semua negara

23595178

Kasus kumulatif terbanyak pada 20 Januari 2022 sebanyak 23.595.178 kasus. Sekarang cek negara mana yang memiliki kasus terbanyak tersebut.

In [214]:
df[
    (df['1/20/22']==23595178) #query negara dengan kasus 23595178 pada 20 Januari 2022 
]

Province/State Country/Region     Lat     Long  1/22/20  1/23/20  1/24/20  \
31            NaN         Brazil -14.235 -51.9253        0        0        0   

    1/25/20  1/26/20  1/27/20  ...   2/10/22   2/11/22   2/12/22   2/13/22  \
31        0        0        0  ...  27135550  27299336  27434286  27492904   

     2/14/22   2/15/22   2/16/22   2/17/22   2/18/22   2/19/22  
31  27552267  27677468  27819996  27940119  28072238  28177367  

[1 rows x 764 columns]

Negara dengan kasus terbanyak tersebut adalah Brazil. Bandingkan dengan Indonesia.

In [215]:
df[
    (df['Country/Region']=='Indonesia') #query di Indonesia
]['1/20/22'] #menampilkan kolom 20 Januari 2022

149    4277644
Name: 1/20/22, dtype: int64

Indonesia memiliki kasus kumulatif sebanyak 4.277.644 pada tanggal yang sama. Kira-kira 1/6 dari Brazil.

## Soal Nomor 2

2. Buatlah fungsi untuk menghitung jumlah kasus harian Covid-19! (Hint: Kamu bisa menggunakan .diff() untuk menghitung perbedaan tiap baris data/menghitung jumlah kasus harian dan .transpose() untuk mengubah kolom menjadi baris atau sebaliknya)

### Import Library dan Data Loading

Library dan dataset yang digunakan masih mengikuti Soal Nomor 1 sehingga tidak di-load ulang.

### Answer

Sebelum membuat fungsi, kita sederhanakan dulu dataset dengan membuat copy dataset, mengubah label menjadi negara, drop beberapa kolom di depan, lalu transpose dataset tersebut.

In [216]:
df_harian = df.copy() #copy dataset ke variabel baru

In [217]:
df_harian.head() #cek lima baris teratas

Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  2/10/22  2/11/22  2/12/22  \
0        0        0        0        0  ...   169448   169940   170152   
1        0        0        0        0  ...   267020   267020   267551   
2        0        0        0        0  ...   260191   260723   261226   
3        0        0        0        0  ...    37074    37140    37140   
4        0        0        0        0  ...    98474    98501    98514   

   2/13/22  2/14/22  2/15/22  2/16/22  2/17/22  2/18/22  2/19/22  
0   170604   171246   171422   171519   171673   171857   171931  
1   268008   268304   268491   268940   269301   269601   269904  
2   261752   262165   262570   262994   263369   263685   263936  
3    37140    37277    37361    37452    37522    37589    37589  
4    98514    98514    98555    98568    98585    98605    98617  

[5 rows x 764 columns]

Mengubah label menggunakan 'Country/Region'

In [218]:
df_harian.set_axis(df_harian['Country/Region'],inplace=True) #mengubah label axis=0 atau baris menggunakan isi kolom Country/Region

In [219]:
df_harian.head() #cek lima baris teratas

Province/State Country/Region       Lat       Long  1/22/20  \
Country/Region                                                               
Afghanistan               NaN    Afghanistan  33.93911  67.709953        0   
Albania                   NaN        Albania  41.15330  20.168300        0   
Algeria                   NaN        Algeria  28.03390   1.659600        0   
Andorra                   NaN        Andorra  42.50630   1.521800        0   
Angola                    NaN         Angola -11.20270  17.873900        0   

                1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  ...  2/10/22  \
Country/Region                                               ...            
Afghanistan           0        0        0        0        0  ...   169448   
Albania               0        0        0        0        0  ...   267020   
Algeria               0        0        0        0        0  ...   260191   
Andorra               0        0        0        0        0  ...    37074   
Angola                0        0        0        0        0  ...    98474   

                2/11/22  2/12/22  2/13/22  2/14/22  2/15/22  2/16/22  2/17/22  \
Country/Region                                                                  
Afghanistan      169940   170152   170604   171246   171422   171519   171673   
Albania          267020   267551   268008   268304   268491   268940   269301   
Algeria          260723   261226   261752   262165   262570   262994   263369   
Andorra           37140    37140    37140    37277    37361    37452    37522   
Angola            98501    98514    98514    98514    98555    98568    98585   

                2/18/22  2/19/22  
Country/Region                    
Afghanistan      171857   171931  
Albania          269601   269904  
Algeria          263685   263936  
Andorra           37589    37589  
Angola            98605    98617  

[5 rows x 764 columns]

Drop beberapa kolom

In [220]:
df_harian.drop(columns=['Province/State','Country/Region','Lat','Long'],inplace=True) #drop kolom yang tidak dibutuhkan untuk perhitungan kasus harian

In [221]:
df_harian.head() #cek lima baris teratas

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  \
Country/Region                                                                  
Afghanistan           0        0        0        0        0        0        0   
Albania               0        0        0        0        0        0        0   
Algeria               0        0        0        0        0        0        0   
Andorra               0        0        0        0        0        0        0   
Angola                0        0        0        0        0        0        0   

                1/29/20  1/30/20  1/31/20  ...  2/10/22  2/11/22  2/12/22  \
Country/Region                             ...                              
Afghanistan           0        0        0  ...   169448   169940   170152   
Albania               0        0        0  ...   267020   267020   267551   
Algeria               0        0        0  ...   260191   260723   261226   
Andorra               0        0        0  ...    37074    37140    37140   
Angola                0        0        0  ...    98474    98501    98514   

                2/13/22  2/14/22  2/15/22  2/16/22  2/17/22  2/18/22  2/19/22  
Country/Region                                                                 
Afghanistan      170604   171246   171422   171519   171673   171857   171931  
Albania          268008   268304   268491   268940   269301   269601   269904  
Algeria          261752   262165   262570   262994   263369   263685   263936  
Andorra           37140    37277    37361    37452    37522    37589    37589  
Angola            98514    98514    98555    98568    98585    98605    98617  

[5 rows x 760 columns]

Transpose dataset

In [222]:
df_harian = df_harian.T #transpose dataset agar negara menjadi kolom dan tanggal menjadi baris

In [223]:
df_harian.head() #cek lima baris teratas

Country/Region  Afghanistan  Albania  Algeria  Andorra  Angola  Antarctica  \
1/22/20                   0        0        0        0       0           0   
1/23/20                   0        0        0        0       0           0   
1/24/20                   0        0        0        0       0           0   
1/25/20                   0        0        0        0       0           0   
1/26/20                   0        0        0        0       0           0   

Country/Region  Antigua and Barbuda  Argentina  Armenia  Australia  ...  \
1/22/20                           0          0        0          0  ...   
1/23/20                           0          0        0          0  ...   
1/24/20                           0          0        0          0  ...   
1/25/20                           0          0        0          0  ...   
1/26/20                           0          0        0          0  ...   

Country/Region  Uruguay  Uzbekistan  Vanuatu  Venezuela  Vietnam  \
1/22/20               0           0        0          0        0   
1/23/20               0           0        0          0        2   
1/24/20               0           0        0          0        2   
1/25/20               0           0        0          0        2   
1/26/20               0           0        0          0        2   

Country/Region  West Bank and Gaza  Winter Olympics 2022  Yemen  Zambia  \
1/22/20                          0                     0      0       0   
1/23/20                          0                     0      0       0   
1/24/20                          0                     0      0       0   
1/25/20                          0                     0      0       0   
1/26/20                          0                     0      0       0   

Country/Region  Zimbabwe  
1/22/20                0  
1/23/20                0  
1/24/20                0  
1/25/20                0  
1/26/20                0  

[5 rows x 284 columns]

Pembuatan fungsi

In [224]:
def kasus_harian(dataframe_input): #deklarasi fungsi dengan nama fungsi kasus_harian, input fungsi berupa dataframe
    dataframe_output = dataframe_input.diff()  #selisih nilai harian dihitung menggunakan .diff() lalu disimpan sebagai variabel dataframe_output
    return dataframe_output #fungsi mengembalikan dataframe_output 

Panggil fungsi dengan dataset yang sudah dibersihkan tadi dan simpan sebagai df_harian yang baru

In [225]:
df_harian = kasus_harian(df_harian) #panggil fungsi dan disimpan sebagai df_harian
df_harian.head() #cek lima baris teratas

Country/Region  Afghanistan  Albania  Algeria  Andorra  Angola  Antarctica  \
1/22/20                 NaN      NaN      NaN      NaN     NaN         NaN   
1/23/20                 0.0      0.0      0.0      0.0     0.0         0.0   
1/24/20                 0.0      0.0      0.0      0.0     0.0         0.0   
1/25/20                 0.0      0.0      0.0      0.0     0.0         0.0   
1/26/20                 0.0      0.0      0.0      0.0     0.0         0.0   

Country/Region  Antigua and Barbuda  Argentina  Armenia  Australia  ...  \
1/22/20                         NaN        NaN      NaN        NaN  ...   
1/23/20                         0.0        0.0      0.0        0.0  ...   
1/24/20                         0.0        0.0      0.0        0.0  ...   
1/25/20                         0.0        0.0      0.0        0.0  ...   
1/26/20                         0.0        0.0      0.0        0.0  ...   

Country/Region  Uruguay  Uzbekistan  Vanuatu  Venezuela  Vietnam  \
1/22/20             NaN         NaN      NaN        NaN      NaN   
1/23/20             0.0         0.0      0.0        0.0      2.0   
1/24/20             0.0         0.0      0.0        0.0      0.0   
1/25/20             0.0         0.0      0.0        0.0      0.0   
1/26/20             0.0         0.0      0.0        0.0      0.0   

Country/Region  West Bank and Gaza  Winter Olympics 2022  Yemen  Zambia  \
1/22/20                        NaN                   NaN    NaN     NaN   
1/23/20                        0.0                   0.0    0.0     0.0   
1/24/20                        0.0                   0.0    0.0     0.0   
1/25/20                        0.0                   0.0    0.0     0.0   
1/26/20                        0.0                   0.0    0.0     0.0   

Country/Region  Zimbabwe  
1/22/20              NaN  
1/23/20              0.0  
1/24/20              0.0  
1/25/20              0.0  
1/26/20              0.0  

[5 rows x 284 columns]

Nilai NaN dapat kita ubah ke 0 karena pada 22 Januari 2020 belum ada kasus sama sekali.

In [226]:
df_harian.fillna(0,inplace=True) #mengubah NaN menjadi 0

In [227]:
df_harian.head() #cek lima baris teratas

Country/Region  Afghanistan  Albania  Algeria  Andorra  Angola  Antarctica  \
1/22/20                 0.0      0.0      0.0      0.0     0.0         0.0   
1/23/20                 0.0      0.0      0.0      0.0     0.0         0.0   
1/24/20                 0.0      0.0      0.0      0.0     0.0         0.0   
1/25/20                 0.0      0.0      0.0      0.0     0.0         0.0   
1/26/20                 0.0      0.0      0.0      0.0     0.0         0.0   

Country/Region  Antigua and Barbuda  Argentina  Armenia  Australia  ...  \
1/22/20                         0.0        0.0      0.0        0.0  ...   
1/23/20                         0.0        0.0      0.0        0.0  ...   
1/24/20                         0.0        0.0      0.0        0.0  ...   
1/25/20                         0.0        0.0      0.0        0.0  ...   
1/26/20                         0.0        0.0      0.0        0.0  ...   

Country/Region  Uruguay  Uzbekistan  Vanuatu  Venezuela  Vietnam  \
1/22/20             0.0         0.0      0.0        0.0      0.0   
1/23/20             0.0         0.0      0.0        0.0      2.0   
1/24/20             0.0         0.0      0.0        0.0      0.0   
1/25/20             0.0         0.0      0.0        0.0      0.0   
1/26/20             0.0         0.0      0.0        0.0      0.0   

Country/Region  West Bank and Gaza  Winter Olympics 2022  Yemen  Zambia  \
1/22/20                        0.0                   0.0    0.0     0.0   
1/23/20                        0.0                   0.0    0.0     0.0   
1/24/20                        0.0                   0.0    0.0     0.0   
1/25/20                        0.0                   0.0    0.0     0.0   
1/26/20                        0.0                   0.0    0.0     0.0   

Country/Region  Zimbabwe  
1/22/20              0.0  
1/23/20              0.0  
1/24/20              0.0  
1/25/20              0.0  
1/26/20              0.0  

[5 rows x 284 columns]

## Soal Nomor 3

3. Pada tanggal berapa puncak tertinggi kasus harian Covid-19 di Indonesia? bandingkan dengan tanggal 20 Januari 2022, apakah mengalami penurunan atau kenaikan jumlah kasus?

### Import Library dan Data Loading

Library dan dataset yang digunakan masih mengikuti Soal Nomor 2 sehingga tidak di-load ulang.

### Answer

Cari kasus tertinggi harian di Indonesia

In [235]:
df_harian[
    (df_harian['Indonesia']==df_harian['Indonesia'].max().max()) #query kolom Indonesia dengan nilai tertinggi
]['Indonesia']

2/16/22    64718.0
Name: Indonesia, dtype: float64

Kasus harian terbanyak terjadi pada tanggal 16 Februari 2022 sebanyak 64.718 kasus/hari. Bandingkan dengan tanggal 20 Januari 2022.

In [230]:
df_harian.loc['1/20/22']['Indonesia'] #query baris 20 Januari 2022 di Indonesia

2116.0

Pada tanggal 20 Januari 2022 hanya terdapat penambahan 2116 kasus. Kenaikan kasus di 16 Februari 2022 sampai 30x lipat dibandingkan tanggal 20 Januari 2022.

## Soal Nomor 4

4. Gunakan for loop dan conditional if untuk membuat sebuah list yang berisikan jumlah kasus harian Covid-19 dengan ketentuan hanya mengambil data Indonesia saja dan jumlah kasus di bawah 10000. Berapa jumlah hari dengan kasus di bawah 10000?

### Import Library dan Data Loading

Library dan dataset yang digunakan masih mengikuti Soal Nomor 2 dan 3 sehingga tidak di-load ulang.

### Answer

Buat for loop terlebih dahulu untuk mengumpulkan elemen dengan nilai di bawah 10.000 dari dataset df_harian

In [231]:
kasus_under_10k = [] #buat list kosong untuk menyimpan output loop
i=0 #variabel untuk indexing
for kasus in df_harian['Indonesia']: #loop untuk tiap nilai di dalam df_harian kolom Indonesia
    if df_harian.iloc[i]['Indonesia']<10000: #condition: Jika kasus di bawah 10000 di Indonesia, maka 
        kasus_under_10k.append(kasus) #masukkan nilai kasus ke variabel kasus_under_10k
    i+=1 #index +1

Output loop disimpan dalam variabel kasus_under_10k. Sekarang kita hanya perlu menghitung jumlah elemen dalam list ini untuk mendapatkan jumlah hari.

In [232]:
len(kasus_under_10k) #cek panjang list

633

Terdapat 633 hari dengan jumlah kasus di bawah 10.000.

Kita juga bisa menggunakan while loop untuk menjawab pertanyaan di atas.

In [233]:
kasus_under_10k_while = [] #buat list kosong untuk menyimpan output loop
i=0 #variabel untuk indexing
while i < len(df_harian['Indonesia']): #selama i kurang dari panjang kolom Indonesia
    if df_harian.iloc[i]['Indonesia']<10000: #condition: Jika kasus di bawah 10000 di Indonesia, maka 
        kasus_under_10k_while.append(kasus) #masukkan nilai kasus ke variabel kasus_under_10k
    i+=1 #index +1

In [234]:
len(kasus_under_10k_while) #cek panjang list

633

# End